working/base-interface-final/feature-specification.md

```diff
@@ -0,0 +1,470 @@
+defaults and keyword choice which may be a useful reference. Unlike that
+proposal, this proposal is non-breaking.
+
+[type modifiers]: https://github.com/dart-lang/language/blob/master/working/type-modifiers/feature-specification.md
+["closed" and "base"]: https://github.com/dart-lang/language/issues/2595
+https://github.com/dart-lang/language/issues/2595
+[leaf proposal]: https://github.com/dart-lang/language/issues/2595
+
+## Motivation
+
+Dart's ethos is to be permissive by default. When you declare a class, it can be
+constructed, subclassed, and even exposes an implicit interface which can be
+implemented—a feature (possibly) unique to Dart. Users generally appreciate this
+flexibility and the power it places in the hands of library consumers.
+
+Why might the author of a class or mixin author want to *remove* capabilities?
+Doesn't that just make the type less useful? The type does end up more
+restricted, but in return, there are more invariants about the type that the
+type author and users can rely on being true. Those invariants may make the type
+easier to understand, maintain, evolve, or even just to use.
+
+Here are some use cases where restricting capabilities may lead to more robust
+software:
+
+### Adding methods
+
+It's a compile-time error to have an `implements` clause and not contain
+definitions of every member in the type that you claim to implement. This is a
+*useful* error because it ensures that any member someone can access on a type
+is actually defined and will succeed. It helps you in case you forget to
+implement something.
+
+But it also means that if a *new* member is added to a class then every single
+class implementing that class's interface now has a new compile error since none
+of them have that member.
+
+This makes it hard to add new members to existing public types in packages.
+Since anyone could be implementing that type's interface, any new member is
+potentially a breaking API change which necessitates a major version bump. In
+practice, many API authors just document "please don't implement this class" and
+then rely on users to not do that.
+
+However, for widely used packages, that polite agreement isn't sufficient.
+Instead, they are simply prevented from adding new members and APIs get frozen
+in time.
+
+If a type disallows being implemented, then it becomes easier to add new members
+without worrying about breaking existing users. If the type also prevents being
+extended, then it's entirely safe to add new members to it. This makes it easier
+to grow and evolve APIs.
+
+### Unintended overriding
+
+Most types contain methods that invoke other methods on `this`, for example:
+
+```dart
+class Account {
+  int _balance = 0;
+
+  bool canWithdraw(int amount) => amount <= _balance;
+
+  bool tryWithdraw(int amount) {
+    if (amount <= 0) throw ArgumentError("Amount must be positive.");
+
+    if (!canWithdraw(amount)) return false;
```

Comment:
This can also be seen as an example of not wanting a virtual invocation.
What if you could write
```dart
 if (!Account.this.canWithdraw(amount)) return false;
```
and get a non-virtual this-call (like a `super`-call, just on this class)?

(But not relevant to this document, just saying that `final` isn't necessarily the *only* solution to this particular problem.)

Reply:
Yeah, there's a whole other much larger discussion to have around non-virtual members that also touches protected, final members (virtual but not overridable), and maybe even overloading.

For now I just want to worry about modifiers on types. Modifiers on members can come later.